jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientResponseListenersTest.java

```diff
@@ -0,0 +1,201 @@
+            .onResponseBegin(listener)
+            .onResponseHeader(listener)
+            .onResponseHeaders(listener)
+            .onResponseContent(listener)
+            .onResponseContentAsync(listener)
+            .onResponseSuccess(listener)
+            .onResponseFailure(listener)
+            .onComplete(listener)
+            .send(listener);
+
+        assertTrue(latch.await(5, TimeUnit.SECONDS));
+        int expectedEventsTriggeredByResponseListeners = 4;
+        int expectedEventsTriggeredBySendListener = 4;
+        int expected = expectedEventsTriggeredByResponseListeners + expectedEventsTriggeredBySendListener;
+        assertEquals(expected, counter.get());
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(ScenarioProvider.class)
+    public void testResponseListenerAddedInResponseListenerEvent(Scenario scenario) throws Exception
+    {
+        start(scenario, new EmptyServerHandler());
+
+        CountDownLatch beginLatch = new CountDownLatch(1);
+        CountDownLatch headersLatch = new CountDownLatch(1);
+        ContentResponse response = client.newRequest("localhost", connector.getLocalPort())
+            .scheme(scenario.getScheme())
+            // Listeners for future events will be notified.
+            .onResponseBegin(r1 -> r1.getRequest().onResponseHeaders(r2 -> beginLatch.countDown()))
+            // Listeners for current or past events won't be notified.
+            .onResponseHeaders(r1 -> r1.getRequest().onResponseHeaders(r2 -> headersLatch.countDown()))
+            .send();
+
+        assertEquals(HttpStatus.OK_200, response.getStatus());
+        assertTrue(beginLatch.await(5, TimeUnit.SECONDS));
+        assertFalse(headersLatch.await(1, TimeUnit.SECONDS));
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(ScenarioProvider.class)
+    public void testConversationResponse(Scenario scenario) throws Exception
+    {
+        start(scenario, new Handler.Abstract()
+        {
+            @Override
+            public boolean handle(Request request, org.eclipse.jetty.server.Response response, Callback callback) throws Exception
+            {
+                switch (Request.getPathInContext(request))
+                {
+                    case "/old" -> org.eclipse.jetty.server.Response.sendRedirect(request, response, callback, "/new");
+                    case "/new" ->
+                    {
+                        response.getHeaders().put("Special", "Value");
+                        Content.Sink.write(response, true, "data", callback);
+                    }
+                    default -> throw new IllegalStateException();
+                }
+                return true;
+            }
+        });
+
+        ContentResponse response = client.newRequest("localhost", connector.getLocalPort())
+            .scheme(scenario.getScheme())
+            .path("/old")
+            .onResponseBegin(r ->
+            {
+                if (r.getStatus() != HttpStatus.OK_200)
```

Comment:
OK, this implies that you are not expecting the listener to be called for intermediate responses?   Could to make this REALLY clear by adding a test `r.getRequest().getPath().equals("/old")`?

Reply:
Added `assertSame()` instead.